Monitor Improvements LUCENE-10422

mogui opened the pull request:
# Description

Change how Monitor manages it's directory, IndexWriter and IndexReader, giving the possibility to
- use a custom Directory implementation
- use a readonly QueryIndex in order to have more Monitor instance on different server reading from the same index (now the index reader is created from the index writer so it is impossible to make a readonly Monitor)

# Solution
I've extended MonitorConfiguration, with a readonly boolean property that governs ho the QueryIndex is created, and I also added a functional property that acts as a Directory Provider

# Tests

I've added tests to confirm two readonly monitor could work as expected.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

nikousai-design committed:
Read  only monitor  implementation

nikousai-design committed:
Format issue and documentation fixes

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/QueryIndex.java:
```diff
@@ -87,8 +81,20 @@
   final Map<IndexReader.CacheKey, QueryTermFilter> termFilters = new HashMap<>();
 
   QueryIndex(MonitorConfiguration config, Presearcher presearcher) throws IOException {
```
Given that QueryIndex is package-private I think we have some leeway to split things up here a bit and make things clearer.  Could you try making QueryIndex itself an interface, and then having read-only and writeable implementations?  We could also move the purge executor from the Monitor into the writeable QueryIndex implementation as it's not necessary to clear the cache at all on a read-only query index.

mogui replied:
HI @romseygeek, thanks for the reply, ok for everything!
 I'll work on that, regarding the interface implementation split, that I agree it's more clear, How you suggest to let user select the implementattion? Keeping the readonly flag in MonitorConfiguration and instantiating the correct on in Monitor, should be enough?

romseygeek replied:
> Keeping the readonly flag in MonitorConfiguration and instantiating the correct on in Monitor, should be enough?

I think so.  Maybe add it as a second parameter to `setDirectoryProvider` as it only really makes sense in combination with a custom Directory?

mogui replied:
Yes, it makes sense.

mogui replied:
I'm trying to wrap my head around it and tried few things, but from what I understand search methods cannot be isolted from cache population, so without a middle abstract class I'll end up with a lot of duplicate code.
I don't like this Interface -> BaseAbstract and two implemtation hierarchy so much. We could skip the interface, but still I don't like so much.

Maybe I'm missing something, but Monitor.search seems very coupled with loading in memory all stored queries in `ConcurrentMap<String, QueryCacheEntry> queries` so I have to replicate all query cache releated code.

Am I wrong?

mogui replied:
Nevermind, I got it or maybe I just think i did :)

Tomorrow I'll update the pull request.

romseygeek reviewed (changes requested):
Thanks for opening this @mogui! I left some comments - I think we can break things up into read-only and writeable classes which should help simplify things a bit.

nikousai-design committed:
Use extisting functional interface and giving it a default

nikousai-design committed:
Moved readonly setting on setDirectoryProvider

nikousai-design committed:
Made QueryIndex an interface implemented by a WritableQueryIndex and ReadonlyQueryIndex classes.

nikousai-design committed:
Moved Purge Executor and listeners from monitor to WritableQueryIndex

mogui committed:
Merge branch 'apache:main' into readonly-monitor

mogui commented:
@romseygeek I've updated with the requested changes

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/Monitor.java:
```diff
@@ -125,14 +108,21 @@ public Monitor(Analyzer analyzer, Presearcher presearcher, MonitorConfiguration
+   * @throws IllegalStateException when Monitor is readonly
    */
   public void addQueryIndexUpdateListener(MonitorUpdateListener listener) {
-    listeners.add(listener);
```
I think we can just make `addListener()` a method on `QueryIndex` and delegate there? And then we don't need the `readOnly` member variable on `Monitor`

mogui replied:
yes better

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,184 @@
+    Directory directory = configuration.getDirectoryProvider().get();
+    this.manager = new SearcherManager(directory, new TermsHashBuilder(termFilters));
+    this.decomposer = configuration.getQueryDecomposer();
+    this.serializer = configuration.getQuerySerializer();
```
AIUI, this implementation doesn't have an in-memory query cache, and re-parses the queries every time we do a match.  I think having a lazy parser is definitely a valid use-case but I think we should decouple it from the notion of a read-only monitor.

mogui replied:
Yes, it all goes to the `MonitorQueryCollector` that relies on the in-memory query  cache, and it is an internal class of WritableQueryIndex.

Are you suggesting to decouple `ReadonlyMonitorQueryCollector` as a lazy query parser,  outside Readonly Monitor ?
Am I getting it right? @romseygeek

romseygeek replied:
So the way the Monitor works at the moment is that it parses all the serialized queries in its QueryIndex on startup, and stores them in an in-memory cache.  This means that when we run a document through the Monitor, once it has identified which candidate queries to run against it we don't need to re-parse them, they are already instantiated in RAM.  The alternative, which I think is what you've implemented here, is to re-parse the query every time we need to run it.  This is perfectly reasonable (in fact it's what elasticsearch's percolator does) but it is a significant change in behaviour so I don't think we should fold it in as part of this ticket.

The Writeable query index needs to have all the tricksy behaviour around re-populating the cache, as we need to remove deleted entries or replace updated entries when queries are added or deleted (cache invalidation is hard, apparently!); the read-only index can just hold everything in a single Map that is populated on startup and never changes.

mogui replied:
Ok it's more clear now.
I can do it that way, but The readonly monitor would need a way to repopulate the cache too, assuming there are other writer that insert and delete on the same index, the readonly would never get the delta untile it gets re-instantiated to populate its Map. 
This way would not be so useful.

mogui replied:
What do you think if I keep all the in-memory cache things along with the purgeExecutor in the abstract QueryIndex class and let the readonly monitor use it too, closing this round of improvements.
Then open another changeset to implement the lazy parsing.

romseygeek replied:
At the moment you wouldn't pick up any changes anyway, because you're not calling `maybeRefresh()` on the SearcherManager so you would always get the same view of the index.  If you want a dynamic view then you'll need  a background refresh thread.

mogui replied:
Ah, good to know! what do you think it's best to do? Make sense what I have proposed?

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,184 @@
+    }
+    BytesRef[] bytesHolder = new BytesRef[1];
+    search(
+        new TermQuery(new Term(WritableQueryIndex.FIELDS.query_id, queryId)),
```
Given that we're re-using a fair amount of stuff from WriteableQueryIndex here, it probably makes more sense to make QueryIndex an abstract class and move the shared code there.

mogui replied:
yes it' something that I thought about, I'll drop the interface for an  abstract class

romseygeek reviewed (changes requested):
Hi @mogui, thanks for your updates.  I left a couple of comments.

mogui commented on lucene/monitor/src/java/org/apache/lucene/monitor/MonitorConfiguration.java:
```diff
@@ -47,16 +49,39 @@ private static IndexWriterConfig defaultIndexWriterConfig() {
     return iwc;
   }
 
+  public Boolean isReadOnly() {
```
sure

nikousai-design committed:
Small improvements
moved common logic to QueryIndex now an abstract class
Removed Lazy Parsing of queries, readonly monitor uses in memory cache just like Writable, but without locking and purge executor.

mogui commented:
OK I've moved things around, making QueryIndex an abstract class sharing common code for both implementations
ReadonlyQueryIndex have the same logic of in-memory laoding of parsed queries, but without locking handling and without purge executor.
I've made it usable by enabling purging cache so it can be done externally.

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,100 @@
+    this.manager = new SearcherManager(directory, new TermsHashBuilder(termFilters));
+    this.decomposer = configuration.getQueryDecomposer();
+    this.serializer = configuration.getQuerySerializer();
+    this.populateQueryCache(serializer, decomposer);
```
So this will still only repopulate the cache when you explicitly call `purgeCache()` on the parent Monitor I think?  That's probably OK but we should document that clearly.

mogui replied:
Yes.
@romseygeek Do you think it could make sense using the purge executor here too?

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,100 @@
+
+  @Override
+  void purgeCache(CachePopulator populator) throws IOException {
+    final ConcurrentMap<String, QueryCacheEntry> newCache = new ConcurrentHashMap<>();
```
I don't think this needs to be a concurrent hash map in this case, as we're not updating it elsewhere.

mogui replied:
True, but then we have to assign iot to `queries` that it is in the abstract class and it is concurrent

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/WritableQueryIndex.java:
```diff
@@ -0,0 +1,253 @@
+  @Override
+  public void close() throws IOException {
+    IOUtils.close(manager, writer, writer.getDirectory());
+    purgeExecutor.shutdown();
```
Let's close the executor first to ensure we don't get spurious exceptions from a cache refresh operating on a closed directory.

romseygeek commented on lucene/monitor/src/test/org/apache/lucene/monitor/TestMonitorReadonly.java:
```diff
@@ -0,0 +1,165 @@
+                new MonitorQuery("query1", query, query.toString(), Collections.emptyMap()));
+          });
+    }
+  }
```
Can you add a test for updating the write monitor? We should see no change in the read monitor until we call purgeCache.

romseygeek reviewed (commented):
Thanks @mogui! I left some more comments.

nikousai-design committed:
Added test to ensure readonly monitor gets changes after purging its cache
Added some docs

mogui commented:
@romseygeek I should have fixed everything, also added few lines of docs to explain read-only behaviour.

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,117 @@
+
+  @Override
+  void purgeCache(CachePopulator populator) throws IOException {
+    manager.maybeRefresh();
```
There's a race here, I think.  Once you've called `manager.maybeRefresh()` then a subsequent call to `search` will return the new searcher, but you may not have updated `this.queries` yet.

mogui replied:
should I call `maybeRefreshBlocking` instead?

romseygeek replied:
I think actually the best solution is to remove the query cache entirely for this impl, which is where you started out - sorry for all the back and forth here.  We can have a background thread that calls maybeRefresh() on the manager to keep up with updates, but all the queries will be read directly from the searcher and parsed as they are executed.  The in-memory cache works when the Monitor in question is handling updates as well, but trying to do that when you have no idea what the changes are between IndexReaders is going to be nasty.

mogui replied:
Ok, I think it was the best solution too, I'll work getting back to that solution.
Don't worry, all the back and forth got me to understand everything better!

mogui replied:
@romseygeek I just realized that main branch from which I started the fork requires java 17, is there a way to make a build that targets Java 11 or do I have to remake the pull request on topo of 9_0_0 branch ?

mogui replied:
@romseygeek in your opinion it makes sense to reuse `purgeFrequency` and `purgeFrequencyUnits`  from `MonitorConfiguration` to configure the refreshing thread or is better a new pairof configurations?

mogui replied:
@romseygeek didi had a chanche to look out the new commits?

nikousai-design committed:
Removed query cache from readonly monitor
live query parsing
added executor to refresh index

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,194 @@
+
+  @Override
+  public void purgeCache() {
+    throw new IllegalStateException("Monitor is readOnly, it has no cache");
```
Let's make this call `manager.maybeRefresh`

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,194 @@
+
+  @Override
+  void purgeCache(CachePopulator populator) {
+    throw new IllegalStateException("Monitor is readOnly, it has no cache");
```
Can this throw `UnsupportedOperationException` instead?

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,194 @@
+
+  @Override
+  public void deleteQueries(List<String> ids) throws IOException {
+    throw new IllegalStateException("Monitor is readOnly cannot delete queries");
```
Same here, this should be UnsupportedOperationException

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,194 @@
+
+  @Override
+  public void clear() throws IOException {
+    throw new IllegalStateException("Monitor is readOnly cannot clear");
```
And here

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,194 @@
+
+  @Override
+  public void addListener(MonitorUpdateListener listener) {
+    throw new IllegalStateException("Monitor is readOnly cannot register listeners");
```
And here :)

romseygeek commented on lucene/monitor/src/java/org/apache/lucene/monitor/ReadonlyQueryIndex.java:
```diff
@@ -0,0 +1,194 @@
+          QueryCacheEntry.decompose(mq, decomposer).stream()
+              .filter(queryCacheEntry -> queryCacheEntry.cacheId.equals(cache_id.utf8ToString()))
+              .findFirst()
+              .orElse(null);
```
Can we throw an exception here instead? `null` is going to cause problems in the collector I think?

romseygeek reviewed (changes requested):
Sorry for the delays @mogui. This is pretty close I think? I asked for a few changes in ReadOnlyQueryCache but the rest is looking good.

You should target `main`, so we need Java 17, yes.

I wonder if we should rename `purgeCache` to something like `refreshCache`, as cacheing only really applies to the writeable Monitor.  Maybe in a followup though, as that's an API change and I'd need to think about deprecations etc.

nikousai-design committed:
Small fixes on throwed exception

nikousai-design committed:
misisng test fix

mogui commented:
@romseygeek fixed !

romseygeek reviewed (changes requested):
I left a couple more comments.  Can you merge in the latest master branch as well, as github reckons there are conflicts now?

nikousai-design committed:
Merge main

nikousai-design committed:
More Fix on Exception types
Moved listeners to abstract class QueryIndex to give listener logic to Readonly too and fix an ugly test

nikousai-design committed:
addListener no more abstract

mogui commented:
@romseygeek I've merged main and restored listeners to the abstarct class, You were right the sleep was very ugly

mogui commented:
@romseygeek sorry to bother you, just a reminder :D

romseygeek commented:
One last thing, can you add a CHANGES.txt entry in the Lucene 9.2.0 New Features section? And then I think we're good to merge! Thanks for all your patience on this.

nikousai-design committed:
Merge branch 'main' into readonly-monitor

* main:
  LUCENE-10473: Make tests a bit faster when running nightly. (#754)
  LUCENE-9905: Fix check in TestPerFieldKnnVectorsFormat#testMergeUsesNewFormat
  LUCENE-9614: Fix rare TestKnnVectorQuery failures
  LUCENE-10472: Fix TestMatchAllDocsQuery#testEarlyTermination (#753)
  LUCENE-10418: Move CHANGES to the correct section.
  LUCENE-10418: Optimize `Query#rewrite` in the non-scoring case. (#672)
  LUCENE-10469: Fix score mode propagation in ConstantScoreQuery. (#750)
  LUCENE-10452, LUCENE-10451: mention hunspell changes in CHANGES.txt
  LUCENE-10452: Hunspell: call checkCanceled less frequently to reduce the overhead (#723)
  Add 9.2.0 section to release notes
  LUCENE-10451 Hunspell: don't perform potentially expensive spellchecking after timeout (#721)
  LUCENE-10463: increment java version to 17 in smoke tester (#748)

nikousai-design committed:
Added Chacnges entry

mogui commented:
Done!
Thanks for Your patience and support @romseygeek !

romseygeek commented on lucene/CHANGES.txt:
```diff
@@ -56,7 +56,7 @@ API Changes
 New Features
 ---------------------
-(No changes)
+* LUCENE-10422: Monitor Improvements: `Monitor` can use a custom `Directory` implementation. `Monitor` can be created with a readonly `QueryIndex` in order to have readonly `Monitor` instances. (Niko Usai)
```
```suggestion
* LUCENE-10422: Monitor Improvements: `Monitor` can use a custom `Directory` 
implementation. `Monitor` can be created with a readonly `QueryIndex` in order to 
have readonly `Monitor` instances. (Niko Usai)
```

romseygeek reviewed (approved):
One small nit, we wrap lines in CHANGES to make it easier to read in text editors

mogui committed:
Update lucene/CHANGES.txt

Co-authored-by: Alan Woodward <romseygeek@gmail.com>

mogui commented:
Ok!

mogui commented:
@romseygeek There is something else  in roadmap or future ideas regarding this package that I can contribute? Or where I can look to contribute more to the project. I'd like to remain involved if I can, it was nice to wotk with.

romseygeek commented:
@mogui there isn't a roadmap for the monitor itself, but please do open issues and/or PRs if you have ideas for other improvements!
